registry/client/auth/authchallenge.go

```diff
@@ -54,12 +58,44 @@ func init() {
 	}
 }
 
-func parseAuthHeader(header http.Header) map[string]authorizationChallenge {
-	challenges := map[string]authorizationChallenge{}
+// Ping pings the provided endpoint to determine its required authorization challenges.
+// If a version header is provided, the versions will be returned.
+func Ping(client *http.Client, endpoint, versionHeader string) ([]Challenge, []string, error) {
```

Comment:
This method is doing too many things. It also a few problematic type issues.

1. It embodies both an authentication challenge and the versions provided by the API. In other contexts, we've formatted the challenge as an error, which can simplify method usage.
2. Two argument parameters are strings and a list of strings is returned, after we've done some parsing.
3. It issues an HTTP request.

We can do all these things, but It might better if we break it down differently:

```
// Check ensures that the request is authorized and provides the supported
// versions, if available. An Challenge error may be returned, indicated that
// the request was authorized and different credentials may be required.
func Check(r *http.Request) ([]APIVersion, error)
```